Add option to delete source files on complete

timothyryanwalsh opened the pull request:
Hi, Artefactual! This is a small suggested addition to transfer.py that might help reduce some of the manual labor still necessary with an automation-tools pipeline - namely, deleting transfer source files after transfers have completed.

sevein commented on transfers/transfer.py:
```diff
@@ -165,6 +166,15 @@ def get_status(am_url, am_user, am_api_key, unit_uuid, unit_type, session, hide_
+                shutil.rmtree(db_unit.path)
+                LOGGER.info('Source files deleted for SIP %s deleted', db_unit.uuid)
+            except OSError as e:
+                LOGGER.info('Error deleting source files: %s', e)
```
@timothyryanwalsh should this be a `warning` instead of `info`?

timothyryanwalsh replied:
Yes, good catch! I've made the change.

ross-spencer commented:
@sevein, in response to your question above, and in relation to some other work I am doing with @timothyryanwalsh I wonder if we should query the storage service and look for status of [`UPLOADED`](https://github.com/artefactual/archivematica-storage-service/blob/3cada6fe040697f4aa684836e4b2eb7d3b19ab50/storage_service/locations/models/package.py#L93). I'm not sure how long the other statuses exist for, but if I look at my local Storage Service, `UPLOADED` seems to be the final resting state. What do you think?

@timothyryanwalsh a question about that though, is: Do you want this PR to satisfy a requirement to delete the transfer source once an AIP has been generated and stored, or somewhere in between those states? - `Transfers.py` does have some functions that connect to the Storage Service. We could also add a function to `AMClient.py` as that continues to grow, to return an individual package's details:

`curl -X GET -H"Authorization: ApiKey test:test" 'http://127.0.0.1:62081/api/v2/file/ad3c0bae-e8a9-448c-bf1a-9730cab80ebe/?format=json`

We could call that from `Transfer.py`. Does that sound okay?

sevein commented:
Sounds like a plan!

timothyryanwalsh commented:
Hi @ross-spencer - thanks for taking a look at this! Ideally, yes, the PR woudl delete the transfer source once the AIP has been stored, so your proposal sounds perfect.

timothyryanwalsh commented:
Hi Ross, things got a little weird with git there for a minute (I'm clearly not used to working with codebases that change while I'm not looking!) but I think this should do the trick?

ross-spencer commented:
@timothyryanwalsh I think this is nearly, there, but we probably need to do some work to merge it. I believe you need to do the following to sync your own master branch (the current position of the branch is a commit back in January so it is missing some of the commits we've used up to this PR specifically):
```
$ git remote add upstream https://github.com/artefactual/automation-tools
$ git fetch upstream
$ git pull upstream master
$ git checkout dev/delete-after-ingest
$ git rebase master
```
I suspect, but can't be sure, that this is where you might have found it weird last time. You'll basically receive warnings now that the branches can't be merged and you need to resolve the conflicts. 

You'll need to resolve each set of conflicts for each of your merges respectively. My process lead to a trail like this:
```
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: Add option to delete source files on complete
Using index info to reconstruct a base tree...
M	transfers/transfer.py
Falling back to patching base and 3-way merge...
Auto-merging transfers/transfer.py
CONFLICT (content): Merge conflict in transfers/transfer.py
error: Failed to merge in the changes.
Patch failed at 0001 Add option to delete source files on complete
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ subl transfers/transfer.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git add transfers/transfer.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git rebase --continue
Applying: Add option to delete source files on complete
Applying: Update README.md
Applying: Write deletion error to LOGGER.warning
Applying: Use AMClient to check SIP status
Using index info to reconstruct a base tree...
M	transfers/transfer.py
.git/rebase-apply/patch:19: trailing whitespace.
def get_status(am_url, am_user, am_api_key, ss_url, ss_user, ss_api_key, unit_uuid, 
.git/rebase-apply/patch:55: trailing whitespace.
        status_info = get_status(am_url, am_user, am_api_key, ss_url, ss_user, ss_api_key, 
warning: 2 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging transfers/transfer.py
CONFLICT (content): Merge conflict in transfers/transfer.py
error: Failed to merge in the changes.
Patch failed at 0004 Use AMClient to check SIP status
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git add transfers/transfer.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git add transfers/transferargs.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git rebase --continue
Applying: Use AMClient to check SIP status
Applying: Rebase on master
Using index info to reconstruct a base tree...
M	transfers/transfer.py
.git/rebase-apply/patch:151: trailing whitespace.
def get_status(am_url, am_user, am_api_key, ss_url, ss_user, ss_api_key, 
.git/rebase-apply/patch:152: trailing whitespace.
                unit_uuid, unit_type, session, hide_on_complete=False, 
.git/rebase-apply/patch:219: trailing whitespace.
            am = AMClient(ss_url=ss_url, ss_user_name=ss_user, 
.git/rebase-apply/patch:733: trailing whitespace.
    parser.add_argument('--delete', action='store_true', 
warning: 4 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging transfers/transferargs.py
CONFLICT (add/add): Merge conflict in transfers/transferargs.py
Auto-merging transfers/transfer.py
CONFLICT (content): Merge conflict in transfers/transfer.py
error: Failed to merge in the changes.
Patch failed at 0005 Rebase on master
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ subl transfers/transfer.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git add transfers/transfer.py
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git add transfers/transferargs.py 
ross-spencer@artefactual:~/git/timwalsh/automation-tools:$ git rebase --continue
Applying: Rebase on master
```
So it does seem a little complex but if you stick with it you'll get closer to a pretty decent rebase. On the last set of resolutions where you're effectively combining your 'rebased commit' with the upstream, you'll need to be careful to pick the commits specific to your PR, rather than the upstream ones.

IDK. It might be an interesting exercise for you, but I wonder, maybe it would be more convenient for you to just to sync your AM-tools repo with the current state of ours, and just apply the delete changes on top for a new PR? That would work too.

Once the change has been made, I'll re-review. It looks to be all there so I don't think it will take too long to get this merged.

timothyryanwalsh commented:
Hi @ross-spencer, thank you for the very helpful feedback! You're probably right that it would have been easier to just make a new PR, but I always appreciate an opportunity to get to know git a bit better :)

I've rebased, fixed the tests to account for the changes, and made flake8 happy so I think everything should be good now. If you'd like to me make any other changes, just let me know.

ross-spencer commented:
Closed via https://github.com/artefactual/automation-tools/pull/79.
